File: sexp_prettify_cli.c
// KiCADv8 Style Prettify S-Expression Formatter (sexp formatter)
// By Brian Khuu, 2024
// This script reformats KiCad-like S-expressions to match a specific formatting style.
// Note: This script modifies formatting only; it does not perform linting or validation.

#include <assert.h>
#include <getopt.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "sexp_prettify.h"

typedef enum styleProfile
{
    STYLE_PROFILE_NONE = 0,
    STYLE_PROFILE_KICAD_STANDARD,
    STYLE_PROFILE_KICAD_COMPACT,
} styleProfile;

const char *compact_list_prefixes_kicad[] = {"pts"};
const int compact_list_prefixes_kicad_size = sizeof(compact_list_prefixes_kicad) / sizeof(compact_list_prefixes_kicad[0]);

const char *shortform_prefixes_kicad[] = {"font", "stroke", "fill", "offset", "rotate", "scale"};
const int shortform_prefixes_kicad_size = sizeof(shortform_prefixes_kicad) / sizeof(shortform_prefixes_kicad[0]);

void putc_handler(char c, void *context_putc) { fputc(c, (FILE *)context_putc); }

void usage(const char *prog_name, bool full)
{
    if (full)
    {
        printf("S-Expression Formatter (Brian Khuu 2024)\n");
        printf("\n");
    }

    printf("Usage:\n");
    printf("  %s [OPTION]... SOURCE [DESTINATION]\n", prog_name);
    if (!full)
    {
        printf("  %s -h          Show Full Help Message\n", prog_name);
    }
    printf("  SOURCE             Source file path. If '-' then use standard stream input\n");
    printf("  DESTINATION        Destination file path. If omitted or '-' then use standard stream output\n");
    printf("\n");

    if (full)
    {
        printf("Options:\n");
        printf("  -h                 Show Help Message\n");
        printf("  -w WRAP_THRESHOLD  Set Wrap Threshold. Must be postive value. (default %d)\n", PRETTIFY_SEXPR_KICAD_DEFAULT_CONSECUTIVE_TOKEN_WRAP_THRESHOLD);
        printf("  -l COMPACT_LIST    Add To Compact List. Must be a string. \n");
        printf("  -k COLUMN_LIMIT    Add To Compact List Column Limit. Must be positive value. (default %d)\n", PRETTIFY_SEXPR_KICAD_DEFAULT_COMPACT_LIST_COLUMN_LIMIT);
        printf("  -s SHORTFORM       Add To Shortform List. Must be a string.\n");
        printf("  -p PROFILE         Predefined Style. (kicad, kicad-compact)\n");
        printf("\n");
        printf("Example:\n");
        printf("  - Use standard input and standard output. Also use KiCAD's standard compact list and shortform setting.\n");
        printf("    %s -l pts -s font -s stroke -s fill -s offset -s rotate -s scale - -\n", prog_name);
    }
}

// Main function
int main(int argc, char **argv)
{
    const char *prog_name = argv[0];

    int wrap_threshold = PRETTIFY_SEXPR_KICAD_DEFAULT_CONSECUTIVE_TOKEN_WRAP_THRESHOLD;

    const char **compact_list_prefixes = NULL;
    int compact_list_prefixes_entries_count = 0;
    int compact_list_prefixes_wrap_threshold = PRETTIFY_SEXPR_KICAD_DEFAULT_COMPACT_LIST_COLUMN_LIMIT;

    const char **shortform_prefixes = NULL;
    int shortform_prefixes_entries_count = 0;

    styleProfile kicad_profile_active = STYLE_PROFILE_NONE;

    while (optind < argc)
    {
        const char c = getopt(argc, argv, "hl:s:p:k:");
        if (c == -1)
        {
            break;
        }

        switch (c)
        {
            case 'h':
            {
                usage(prog_name, true);
                return EXIT_SUCCESS;
            }

            case 'l':
            {
                if (!compact_list_prefixes)
                {
                    compact_list_prefixes = malloc(sizeof(compact_list_prefixes));
                }
                else
                {
                    compact_list_prefixes = realloc(compact_list_prefixes, sizeof(compact_list_prefixes) * (compact_list_prefixes_entries_count + 1));
                }
                compact_list_prefixes[compact_list_prefixes_entries_count] = optarg;
                compact_list_prefixes_entries_count++;
                break;
            }

            case 's':
            {
                if (!shortform_prefixes)
                {
                    shortform_prefixes = malloc(sizeof(shortform_prefixes));
                }
                else
                {
                    shortform_prefixes = realloc(shortform_prefixes, sizeof(shortform_prefixes) * (shortform_prefixes_entries_count + 1));
                }
                shortform_prefixes[shortform_prefixes_entries_count] = optarg;
                shortform_prefixes_entries_count++;
                break;
            }

            case 'w':
            {
                const int value = atoi(optarg);

                if (value < 0)
                {
                    usage(prog_name, false);
                    return EXIT_FAILURE;
                }

                wrap_threshold = value;
                break;
            }

            case 'k':
            {
                const int value = atoi(optarg);

                if (value < 0)
                {
                    usage(prog_name, false);
                    return EXIT_FAILURE;
                }

                compact_list_prefixes_wrap_threshold = value;
                break;
            }

            case 'p':
            {
                if (strcmp("kicad", optarg) == 0)
                {
                    kicad_profile_active = STYLE_PROFILE_KICAD_STANDARD;
                }
                else if (strcmp("kicad-compact", optarg) == 0)
                {
                    kicad_profile_active = STYLE_PROFILE_KICAD_COMPACT;
                }

                if (kicad_profile_active == STYLE_PROFILE_NONE)
                {
                    fprintf(stderr, "Must be either 'kicad' or 'kicad-compact'");
                    usage(prog_name, false);
                    return EXIT_FAILURE;
                }

                if (compact_list_prefixes_entries_count)
                {
                    free(compact_list_prefixes);
                    compact_list_prefixes = NULL;
                    compact_list_prefixes_entries_count = 0;
                }

                if (shortform_prefixes_entries_count)
                {
                    free(shortform_prefixes);
                    shortform_prefixes = NULL;
                    shortform_prefixes_entries_count = 0;
                }

                if (kicad_profile_active == STYLE_PROFILE_KICAD_STANDARD || kicad_profile_active == STYLE_PROFILE_KICAD_COMPACT)
                {
                    compact_list_prefixes = malloc(sizeof(compact_list_prefixes));
                    for (int i = 0; i < compact_list_prefixes_kicad_size; i++)
                    {
                        compact_list_prefixes = realloc(compact_list_prefixes, sizeof(compact_list_prefixes) * (compact_list_prefixes_entries_count + 1));
                        compact_list_prefixes[compact_list_prefixes_entries_count] = compact_list_prefixes_kicad[i];
                        compact_list_prefixes_entries_count++;
                    }
                }

                if (kicad_profile_active == STYLE_PROFILE_KICAD_COMPACT)
                {
                    shortform_prefixes = malloc(sizeof(shortform_prefixes));
                    for (int i = 0; i < shortform_prefixes_kicad_size; i++)
                    {
                        shortform_prefixes = realloc(shortform_prefixes, sizeof(shortform_prefixes) * (shortform_prefixes_entries_count + 1));
                        shortform_prefixes[shortform_prefixes_entries_count] = shortform_prefixes_kicad[i];
                        shortform_prefixes_entries_count++;
                    }
                }

                break;
            }

            case '?':
            {
                usage(prog_name, false);
                return EXIT_FAILURE;
            }

            default:
            {
                usage(prog_name, false);
                return EXIT_FAILURE;
            }
        }
    }

    // Get fixed arguments
    const char *src_path = NULL;
    const char *dst_path = NULL;

    if (optind < argc)
    {
        src_path = argv[optind++];
    }

    if (optind < argc)
    {
        dst_path = argv[optind++];
    }

    if (!src_path)
    {
        fprintf(stderr, "Source Path Missing\n");
        usage(prog_name, true);
        return EXIT_SUCCESS;
    }

    // Get File Descriptor
    FILE *src_file = stdin;
    if (src_path && strcmp(src_path, "-") != 0)
    {
        src_file = fopen(src_path, "r");
        if (!src_file)
        {
            perror("Error opening source file");
            return EXIT_FAILURE;
        }
    }

    // Open the destination file else default to standard output
    FILE *dst_file = stdout;
    if (dst_path && strcmp(dst_path, "-") != 0)
    {
        dst_file = fopen(dst_path, "wb");
        if (!dst_file)
        {
            perror("Error opening destination file");
            fclose(src_file);
            return EXIT_FAILURE;
        }
    }

    // Initialise and sanity check
    struct PrettifySExprState state = {0};

    assert(sexp_prettify_init(&state, PRETTIFY_SEXPR_KICAD_DEFAULT_INDENT_CHAR, PRETTIFY_SEXPR_KICAD_DEFAULT_INDENT_SIZE, wrap_threshold));

    if (compact_list_prefixes_entries_count > 0)
    {
        assert(sexp_prettify_compact_list_set(&state, compact_list_prefixes, compact_list_prefixes_entries_count, compact_list_prefixes_wrap_threshold));
    }

    if (shortform_prefixes_entries_count > 0)
    {
        assert(sexp_prettify_shortform_set(&state, shortform_prefixes, shortform_prefixes_entries_count));
    }

    // Process Source Files
    char src_char;
    while ((src_char = fgetc(src_file)) != EOF)
    {
        sexp_prettify(&state, src_char, &putc_handler, dst_file);
    }

    // Wrapup and Cleanup
    fclose(src_file);
    if (dst_file != stdout)
    {
        fclose(dst_file);
    }

    return EXIT_SUCCESS;
}

File: sexp_prettify_cli.py
#!/usr/bin/env python3
import argparse
import sys

def prettify(source, 
             compact_save = False, 
             indent_char = '\t',
             indent_size = 1,
             consecutive_token_wrap_threshold = 72, 
             compact_list_prefixes = {"pts"}, 
             compact_list_column_limit = 99,
             shortform_prefixes = {"font", "stroke", "fill", "offset", "rotate", "scale"}):
    """
    Reformats KiCad-like S-expressions to match a specific formatting style.

    Args:
        source (str): The source S-expression string.
        compact_save (bool): If True, enables compact mode formatting.
    
    Returns:
        str: The formatted S-expression.
    """

    # State tracking
    formatted = []
    list_depth = 0
    column = 0
    previous_non_space_output = ''
    in_quote = False
    escape_next_char = False
    singular_element = False
    space_pending = False
    wrapped_list = False
    scanning_for_prefix = False
    prefix_token = ''
    compact_list_mode = False
    compact_list_indent = 0
    shortform_mode = False
    shortform_indent = 0

    for c in source:

        # Parse quoted strings
        if c == '"' or in_quote:
            if space_pending:
                formatted.append(' ')
                column += 1
                space_pending = False

            if escape_next_char:
                escape_next_char = False
            elif c == '\\':
                escape_next_char = True
            elif c == '"':
                in_quote = not in_quote

            formatted.append(c)
            column += 1
            previous_non_space_output = c
            continue

        # Parse spaces and newlines
        if c.isspace():
            space_pending = True

            if scanning_for_prefix:
                if prefix_token in compact_list_prefixes:
                    compact_list_mode = True
                    compact_list_indent = list_depth
                elif compact_save and prefix_token in shortform_prefixes:
                    shortform_mode = True
                    shortform_indent = list_depth

                scanning_for_prefix = False
            continue

        # Parse opening parentheses
        if c == '(':
            space_pending = False
            if compact_list_mode:
                # In fixed listDepth, visually compact mode
                if column < compact_list_column_limit and previous_non_space_output == ')' or compact_list_column_limit == 0:
                    # Is a consecutive list and still within column limit (or column limit disabled)
                    formatted.append(' ')
                    column += 1
                else:
                    # List is either beyond column limit or not after another list move this list to the next line
                    formatted.append('\n')
                    formatted.append(indent_char * compact_list_indent * indent_size)
                    column = compact_list_indent * indent_size
            elif shortform_mode:
                # In one liner mode
                formatted.append(' ')
                column += 1
            else:
                # Start scanning for prefix for special list handling
                scanning_for_prefix = True
                prefix_token = ''
                if list_depth > 0:
                    # Print next line depth
                    formatted.append('\n')
                    formatted.append(indent_char * list_depth * indent_size)
                    column = list_depth * indent_size

            singular_element = True
            list_depth += 1

            formatted.append('(')
            column += 1

            previous_non_space_output = '('
            continue

        # Parse closing parentheses
        if c == ')':
            current_shortform_mode = shortform_mode
            space_pending = False
            scanning_for_prefix = False

            if list_depth > 0:
                list_depth -= 1

            if compact_list_mode and list_depth < compact_list_indent:
                compact_list_mode = False
            if shortform_mode and list_depth < shortform_indent:
                shortform_mode = False

            if wrapped_list:
                # This was a list with wrapped tokens so is already indented
                formatted.append('\n')
                formatted.append(indent_char * list_depth * indent_size)
                column = list_depth * indent_size
                if singular_element:
                    singular_element = False
                wrapped_list = False
            else:
                # Normal List
                if singular_element:
                    singular_element = False
                elif not current_shortform_mode:
                    formatted.append('\n')
                    formatted.append(indent_char * list_depth * indent_size)
                    column = list_depth * indent_size

            formatted.append(')')
            column += 1

            if list_depth <= 0:
                formatted.append('\n')
                column = 0

            previous_non_space_output = ')'
            continue

        # Parse characters
        if c != '\0':
            if previous_non_space_output == ')' and not shortform_mode:
                # Is Bare token after a list that should be on next line
                # Dev Note: In KiCAD this may indicate a flag bug
                formatted.append('\n')
                formatted.append(indent_char * list_depth * indent_size)
                column = list_depth * indent_size
                space_pending = False
            elif space_pending and not shortform_mode and not compact_list_mode and column >= consecutive_token_wrap_threshold:
                # Token is above wrap threshold. Move token to next line (If token wrap threshold is zero then this feature is disabled)
                wrapped_list = True
                formatted.append('\n')
                formatted.append(indent_char * list_depth * indent_size)
                column = list_depth * indent_size
                space_pending = False
            elif space_pending and previous_non_space_output != '(':
                formatted.append(' ')
                column += 1
                space_pending = False

            # Add to prefix scanning buffer if scanning for special list handling detection
            if scanning_for_prefix:
                prefix_token += c

            # Add character to list
            formatted.append(c)
            column += 1
            previous_non_space_output = c
            continue

    return ''.join(formatted)


def main():
    parser = argparse.ArgumentParser(
        description="KiCad S-Expression Formatter"
    )
    parser.add_argument("src", help="Source file path ('-' for stdin)")
    parser.add_argument("dst", nargs="?", default="-", help="Destination file path ('-' for stdout)")
    parser.add_argument("-c", action="store_true", help="Use compact mode")
    parser.add_argument("-p", help="Predefined Style. (kicad, kicad-compact)")
    args = parser.parse_args()

    # Read input
    if args.src == "-":
        source = sys.stdin.read()
    else:
        with open(args.src, 'r', encoding='utf-8') as f:
            source = f.read()

    # Process formatting
    result = prettify(source = source, compact_save = args.c or args.p == "kicad-compact")

    # Write output
    if args.dst == "-":
        sys.stdout.write(result)
    else:
        with open(args.dst, 'w', encoding='utf-8') as f:
            f.write(result)


if __name__ == "__main__":
    main()
